pkg/oci/oci.go

```diff
@@ -0,0 +1,316 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/distribution/reference"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry/remote"
+	oras_auth "oras.land/oras-go/v2/registry/remote/auth"
+)
+
+type AuthOptions struct {
+	AuthFile string
+	Insecure bool
+}
+
+var (
+	defaultOciStore = "/var/lib/ig/oci-store"
+	DefaultAuthFile = "/var/lib/ig/config.json"
+)
+
+func GetLocalOciStore() (*oci.Store, error) {
+	if err := os.MkdirAll(filepath.Dir(defaultOciStore), 0o710); err != nil {
+		return nil, err
+	}
+	return oci.New(defaultOciStore)
+}
+
+func GetMemoryStore() *memory.Store {
+	return memory.New()
+}
+
+func PullIfNotExist(imageStore oras.Target, authOpts *AuthOptions, image string) error {
+	_, err := imageStore.Resolve(context.TODO(), image)
+	if err == nil {
+		return nil
+	}
+	if !errors.Is(err, errdef.ErrNotFound) {
+		return fmt.Errorf("resolve image %q: %w", image, err)
+	}
+
+	repo, err := NewRepository(image, authOpts)
+	if err != nil {
+		return fmt.Errorf("create remote repository: %w", err)
+	}
+	_, err = oras.Copy(context.TODO(), repo, image, imageStore, image, oras.DefaultCopyOptions)
+	if err != nil {
+		return fmt.Errorf("download to local repository: %w", err)
+	}
+	return nil
+}
+
+func GetImageListDescriptor(imageStore oras.ReadOnlyTarget, image string) (ocispec.Index, error) {
+	imageListDescriptor, err := imageStore.Resolve(context.TODO(), image)
+	if err != nil {
+		return ocispec.Index{}, fmt.Errorf("resolve image %q: %w", image, err)
+	}
+	if imageListDescriptor.MediaType != ocispec.MediaTypeImageIndex {
+		return ocispec.Index{}, fmt.Errorf("image %q is not an image index", image)
+	}
+
+	reader, err := imageStore.Fetch(context.Background(), imageListDescriptor)
+	if err != nil {
+		return ocispec.Index{}, fmt.Errorf("fetch image index: %w", err)
+	}
+	defer reader.Close()
+	bytes, err := io.ReadAll(reader)
+	if err != nil {
+		return ocispec.Index{}, fmt.Errorf("read image index: %w", err)
+	}
+	var index ocispec.Index
+	err = json.Unmarshal(bytes, &index)
```

Comment:
This could be handled more stream-like (also reducing code / error checks):

```suggestion
	var index ocispec.Index
	err = json.NewDecoder(reader).Decode(&index)
```

Reply:
TIL, thanks!